### PR TITLE
Statically linked Python interpreter

### LINKED ISSUE #253: Please installation does not respect the env python (specifically when looking for `cffi`)
I'm not sure how necessary that is, but I'm forced to install `python-pip` on Linux, then pip install cffi using that system pip. Everything else in please respects the path configuration in a `plzconfig`, which allows me to preface the system python places with pyenv shim places, and so get a pyenv-controlled Python installation to take care of all python things. Is it possible to get the bootstrapping to either respect the path entries in a `plzconfig`, or be configurable some other way?

### PR DESCRIPTION
**Pull request by peterebden**
Another mad interpreter scheme - I am literally† snowed in today so clearly there is nothing better to do than be working on the build system...

Background: see discussion in #253. I mentioned there that ideally we'd have less dependency on the system Python, so this moves in that direction. It statically links a Python into the binary to run our parser rather than dynamically loading a thing at runtime which then goes off and grabs other system bits and pieces. Somewhat surprisingly that actually works quite well and caused rather fewer unforeseen problems than expected.

Top-level notes:
- Do not need anything installed to run plz after download, other than standard libraries, but we can assume their existence; the set we're using are within the Python manylinux profile (apart from libz, but everyone that has Python has that anyway...)
- Currently using python2 :( 
  More detailed notes in the PR but it's significantly (~30%) faster than python3 and PyPy isn't an option right now. Can be fairly easily revisited later though.
- We use very little of the stdlib. What we do use I've mostly vendorised to cut down on what we have to ship - e.g. we need the lightest bits of `os`, but to import the whole module normally needs a lot of other things.
- Can still use any of python2, python3 and PyPy at bootstrap time. I'm leaving the old dynamic loading scheme in to facilitate that, but also it's not a huge cost really. Have ditched the portable PyPy download though.
- Download size increases by about 2MB which seems reasonable, and we probably save that much anyway by removing "you must have installed cffi" from about five hundred places in the documentation :)

Not tested:
- Updating. Have attempted to simplify the download scheme as much as possible (e.g. not adding subdirectories) so old versions should be able to update OK.
- OSX. I'm sure CircleCI will tell me the error of my ways shortly.

@natpen may well be interested in this too since it resolves #253 

† maybe more figuratively than literally...

### CODE REVIEW

File: tools/images/ubuntu/Dockerfile [hunk 1]
```diff
@@ -2,29 +2,26 @@ FROM ubuntu:xenial
 MAINTAINER peter.ebden@gmail.com
 
 # Basic dependencies, Python and Java
-RUN apt-get update && apt-get install -y python2.7 python3.5 python-pip python3-pip openjdk-8-jdk curl unzip git locales
-RUN pip install cffi
+RUN apt-get update
+RUN apt-get install -y python2.7 python3.5 python3-pip libpython2.7-dev openjdk-8-jdk-headless curl unzip git locales pkg-config zlib1g-dev
```
Comment (dimpavloff) [inline tools/images/ubuntu/Dockerfile]: the `apt-get update` should be on the same line as the `apt-get install` or you'll have docker layer caching issues

File: tools/images/ubuntu/Dockerfile [hunk 2]
```diff
@@ -2,29 +2,26 @@ FROM ubuntu:xenial
...
-RUN ln -s /usr/local/pypy/bin/libpypy-c.so /usr/lib
-
 # C++
-RUN apt-get install -y pkg-config clang
-RUN curl -fsSL https://github.com/unittest-cpp/unittest-cpp/releases/download/v2.0.0/unittest-cpp-2.0.0.tar.gz | tar -xzC /tmp
-RUN cd /tmp/unittest-cpp-2.0.0 && ./configure --prefix=/usr && make -j4 && make install
-RUN rm -r /tmp/unittest-cpp-2.0.0
+# Ubuntu's packaged version of UnitTest++ is older than we'd like and has some
+# idiosyncratic naming (unittest++ vs. UnitTest++ etc).
+RUN apt-get install -y clang
```
Comment (dimpavloff) [inline tools/images/ubuntu/Dockerfile]: same here, this should be prefixed by `apt-get update`

File: build_defs/engines.build_defs [hunk 3]
```diff
@@ -7,3 +7,55 @@ ENGINE_DEPENDENCIES = {
...
+
+PARSER_PKG_CONFIG = _PARSER_PKG_CONFIG[CONFIG.PARSER_ENGINE]
+PARSER_LIB_FILENAME = _PARSER_FILENAMES[CONFIG.PARSER_ENGINE]
+PARSER_TOOL = _PARSER_BINARIES[CONFIG.PARSER_ENGINE]
+PARSER_LIB_NAME = PARSER_TOOL  # Happens to be the same for now.
+
+# These are really awkward; OSX's linker doesn't support the -Bstatic/-Bdynamic flags
+# and so it's unreasonably difficult to persuade it to link this statically.
+if CONFIG.OS == 'darwin':
+    PARSER_LINKER_FLAGS = ('`pkg-config --libs-only-L python-2.7 | sed "s/^-L//g"`/%s/config/%s -ldl' %
```
[the remark this answers is not in the document]
Comment (peterebden) [reply build_defs/engines.build_defs]: nope, it shouldn't be. Missed that one.

I might do a little fiddle after this to make pkg-config a proper tool and hence issue some slightly more detailed error messages when we can't find it.

File: src/BUILD_ [hunk 4]
```diff
@@ -27,15 +28,31 @@ go_binary(
...
-        # have them in the same directory as the Please binary itself.
-        binary = True,
-    )
+# Needed at runtime for cffi to do it's thing.
+genrule(
+    name = 'cffi_backend',
+    srcs = ['//third_party/python:cffi'],
+    outs = ['_cffi_backend.so'],
+    binary = True,
+    cmd = 'cp third_party/python/cffi/_cffi_backend.so .',
```
[the remark this answers is not in the document]
Comment (peterebden) [reply src/BUILD_]: I don't think so? We control timestamps within archives for determinism but generally don't for the output files themselves. It's not taken into account by any of the rest of the system though.

File: src/parse/interpreter.c [hunk 5]
```diff
@@ -2,6 +2,14 @@
...
+#ifdef BUILD_STATIC_INTERPRETER
+#include "src/parse/cffi/defs.h"
+
+extern int Py_NoSiteFlag;
+#endif
+
+typedef int (*RegisterFunction)(char*, char*, void*);
+
 // Since we dlsym() the callbacks out of the parser .so, we have variables for them as
 // well as extern definitions which cffi uses. The two must match, of course.
```
[the remark this answers is not in the document]
Comment (peterebden) [reply src/parse/interpreter.c]: It's a little inaccurate but still relevant (that these definitions are used and must match the real functions in type. Have updated a bit.

File: tools/images/ubuntu/Dockerfile (hunk 1 above)
Comment (peterebden) [reply tools/images/ubuntu/Dockerfile]: Done, I've collapsed all the apt stuff into one step to save on work.

**Comment by peterebden** [comment]
Thanks Dimitar!

For some reason I can't find how to reply to your review in one go; I've done all the smaller comments but haven't replied to each to cut down on messages.

File: src/BUILD_ (hunk 4 above)
[the remark this answers is not in the document]
Comment (dimpavloff) [reply src/BUILD_]: Ah, right, I wasn't clear on that.

**Pull request merged**